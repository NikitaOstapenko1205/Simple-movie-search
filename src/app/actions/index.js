import axios from 'axios';
import { createAction } from 'redux-actions';

export const fetchMoviesRequest = createAction('MOVIES_FETCH_REQUEST');
export const fetchMoviesSuccess = createAction('MOVIES_FETCH_SUCCESS');
export const fetchSingleMovieSuccess = createAction('MOVIE_SINGLE_FETCH_SUCCESS');
export const fetchMoviesFailure = createAction('MOVIES_FETCH_FAILURE');
export const updateSearchField = createAction('MOVIES_SEARCH_UPDATE');
export const closeModalWindow = createAction('MODAL_WINDOW_CLOSE');

export const fetchMovies = ({ search }) => async dispatch => {
  dispatch(fetchMoviesRequest());
  try {
    const url = `http://www.omdbapi.com/?apikey=9e7bd865&s=${search}&type=movie`;
    const response = await axios.get(url);
    dispatch(fetchMoviesSuccess({ movies: response.data.Search }));
  } catch (e) {
    dispatch(fetchMoviesFailure());
    throw e;
  }
};

export const fetchSingleMovie = ({ id }) => async dispatch => {
  dispatch(fetchMoviesRequest());
  try {
    const url = `http://www.omdbapi.com/?apikey=9e7bd865&i=${id}&plot=full`;
    const response = await axios.get(url);
    dispatch(fetchSingleMovieSuccess({ movie: response.data }));
  } catch (e) {
    dispatch(fetchMoviesFailure());
    throw e;
  }
};
