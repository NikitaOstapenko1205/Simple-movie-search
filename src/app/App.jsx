import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as actions from './actions';

const mapStateToProps = state => {
  return state;
};

const actionCreators = {
  updateSearchField: actions.updateSearchField,
  fetchMovies: actions.fetchMovies,
  fetchSingleMovie: actions.fetchSingleMovie,
  closeModalWindow: actions.closeModalWindow
};

class App extends React.Component {
  static propTypes = {
    fetchMovies: PropTypes.func,
    updateSearchField: PropTypes.func,
    fetchSingleMovie: PropTypes.func,
    closeModalWindow: PropTypes.func,
    search: PropTypes.string,
    moviesFetchingState: PropTypes.string,
    moviesState: PropTypes.object
  };

  updateSearch = e => {
    const { updateSearchField } = this.props;
    updateSearchField({ search: e.target.value });
  };

  findMovies = e => {
    e.preventDefault();
    const { fetchMovies, search } = this.props;
    try {
      fetchMovies({ search });
    } catch (error) {
      throw error;
    }
  };

  findMovie = id => e => {
    e.preventDefault();
    const { fetchSingleMovie } = this.props;
    try {
      fetchSingleMovie({ id });
    } catch (error) {
      throw error;
    }
  };

  addMovies = movies => {
    const htmledMovies = (
      <div className="grid">
        {movies.map(mv => (
          <div className="grid-item" key={mv.imdbID}>
            <h2 className="grid-item-title">{mv.Title}</h2>
            <button type="button" onClick={this.findMovie(mv.imdbID)}>
              <img
                className="grid-item-image"
                alt={mv.Title}
                title="Click for more information"
                src={mv.Poster === 'N/A' ? './img/no_image.png' : mv.Poster}
              />
            </button>
            <div className="grid-item-year">{mv.Year}</div>
            <a
              className="grid-item-imdb"
              rel="noreferrer noopener"
              target="_blank"
              href={`https://www.imdb.com/title/${mv.imdbID}/`}
            >
              IMDB
            </a>
          </div>
        ))}
      </div>
    );
    return htmledMovies;
  };

  closeModal = e => {
    e.preventDefault();
    const { closeModalWindow } = this.props;
    closeModalWindow();
  };

  showMovieModal = () => {
    const {
      moviesState: { movie }
    } = this.props;
    return (
      <div className="movie-modal">
        <button className="movie-modal-close" onClick={this.closeModal} type="button">
          âœ–
        </button>
        <img
          className="movie-modal-image"
          alt={movie.Title}
          title={movie.Title}
          src={movie.Poster === 'N/A' ? './img/no_image.png' : movie.Poster}
        />
        <h2 className="movie-modal-title">{movie.Title}</h2>
        <div className="movie-modal-info movie-modal-date">
          <span>Released: </span>
          {movie.Released}
        </div>
        <div className="movie-modal-info movie-modal-plot">
          <span>Plot: </span>
          {movie.Plot}
        </div>
        <div className="movie-modal-info movie-modal-director">
          <span>Director: </span>
          {movie.Director}
        </div>
        <div className="movie-modal-info movie-modal-actors">
          <span>Actors: </span>
          {movie.Actors}
        </div>
        <div className="movie-modal-info movie-modal-metascore">
          <span>Metascore: </span>
          {movie.Metascore}
        </div>
        <div className="movie-modal-info movie-modal-imdbrating">
          <span>imdbRating: </span>
          {movie.imdbRating}
        </div>
      </div>
    );
  };

  render() {
    const { search, moviesFetchingState, moviesState } = this.props;
    return (
      <div>
        <form onSubmit={this.findMovies}>
          <input
            placeholder="Write the name of the film or part of it"
            type="text"
            required
            value={search}
            onChange={this.updateSearch}
          />
          <button disabled={moviesFetchingState === 'required'} type="submit" />
        </form>
        {moviesState.movies.length > 0 && this.addMovies(moviesState.movies)}
        {!moviesState.moviesFinded && (
          <h2 className="movies-error">Sorry... We found nothing, try again!</h2>
        )}
        {Object.keys(moviesState.movie).length > 0 && this.showMovieModal()}
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  actionCreators
)(App);
