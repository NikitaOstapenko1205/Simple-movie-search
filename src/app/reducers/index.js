import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions';

const moviesFetchingState = handleActions(
  {
    [actions.fetchMoviesRequest]() {
      return 'requested';
    },
    [actions.fetchMoviesFailure]() {
      return 'failed';
    },
    [actions.fetchMoviesSuccess]() {
      return 'finished';
    }
  },
  'none'
);

const moviesState = handleActions(
  {
    [actions.fetchMoviesSuccess](
      state,
      {
        payload: { movies }
      }
    ) {
      const newState = movies
        ? { moviesFinded: true, movies, movie: {} }
        : { moviesFinded: false, movies: [], movie: {} };
      return newState;
    },
    [actions.fetchSingleMovieSuccess](
      state,
      {
        payload: { movie }
      }
    ) {
      return { ...state, movie };
    },
    [actions.closeModalWindow](state) {
      return { ...state, movie: {} };
    }
  },
  { moviesFinded: true, movies: [], movie: {} }
);

const search = handleActions(
  {
    [actions.fetchMoviesSuccess]() {
      return '';
    },
    [actions.updateSearchField](state, { payload }) {
      return payload.search;
    }
  },
  ''
);

export default combineReducers({
  moviesFetchingState,
  moviesState,
  search
});
