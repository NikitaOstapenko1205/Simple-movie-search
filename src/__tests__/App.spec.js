import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../app/App.jsx';
import reducers from '../app/reducers';
import * as actions from '../app/actions';
import store from '../index';

describe('Reducers tests', () => {
  it('Check updateSearchField reducer', () => {
    expect(reducers('', actions.updateSearchField({ search: 'lol' })).search).toBe('lol');
  });

  it('Check on fail fetchMovies reducer', () => {
    expect(reducers('', actions.fetchMoviesFailure()).moviesFetchingState).toBe('failed');
  });

  it('Check on request fethMovies reducer', () => {
    expect(reducers('', actions.fetchMoviesRequest()).moviesFetchingState).toBe('requested');
  });

  it('Check on success fethMovies reducer', () => {
    expect(reducers('', actions.fetchMoviesSuccess({ movies: [] })).moviesFetchingState).toBe(
      'finished'
    );
  });
});

describe('Component snapshot tests', () => {
  const myComponent = mount(<App store={store} />);

  it('Drawing initial state', () => {
    expect(myComponent.find('input').instance().value).toBe('');
    expect(myComponent.find('input').html()).toMatchSnapshot();
  });

  it('Change search input', () => {
    myComponent.find('input').simulate('change', { target: { value: 'lol' } });
    expect(myComponent.find('input').instance().value).toBe('lol');
    expect(myComponent.find('input').html()).toMatchSnapshot();
  });
});
