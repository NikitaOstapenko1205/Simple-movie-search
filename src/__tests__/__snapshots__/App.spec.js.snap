// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot tests Change search input 1`] = `"<input placeholder=\\"Write the name of the film or part of it\\" type=\\"text\\" required=\\"\\" value=\\"lol\\">"`;

exports[`Component snapshot tests Drawing initial state 1`] = `"<input placeholder=\\"Write the name of the film or part of it\\" type=\\"text\\" required=\\"\\" value=\\"\\">"`;
